core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go

```diff
@@ -0,0 +1,146 @@
+// only if the peer is a member of the collection or owns the implicit collection. However, this function
+// adopts a simple implementation that always adds private data namespace for a collection without checking
+// peer membership/ownership. As a result, it may return more private data namespaces that what should be.
+// However, it will not cause any inconsistent issue because the caller in statecouchdb will check if any
+// existing database matches the namespace and filter out all extra namespaces since no databases match them.
+// Checking peer membership is complicated because it requires retrieving all the collection configs from
+// the collection config store. Because this is a temporary function needed to retroactively build namespaces
+// when upgrading v2.0/2.1 peers to a newer v2.x version and because returning extra private data namespaces
+// does not cause inconsistence, it makes sense to use the simple implementation.
+func (p *NamespaceProvider) GetStateDBNamespaces(channelName string, vdb statedb.VersionedDB) ([]string, error) {
+	mspIDs, err := p.GetAllMSPIDs()
+	if err != nil {
+		return nil, err
+	}
+
+	chaincodesInfo, err := p.AllChaincodesInfo(channelName, &simpleQueryExecutor{vdb})
+	if err != nil {
+		return nil, err
+	}
+
+	// iterate each chaincode, add namespaces for chaincodes and explicit collections
+	retNamespaces := []string{}
+	for _, ccInfo := range chaincodesInfo {
+		ccName := ccInfo.Name
+		retNamespaces = append(retNamespaces, ccName)
+		if ccInfo.ExplicitCollectionConfigPkg == nil {
+			continue
+		}
+		for _, config := range ccInfo.ExplicitCollectionConfigPkg.Config {
+			collConfig := config.GetStaticCollectionConfig()
+			if collConfig != nil {
+				retNamespaces = append(retNamespaces, deriveHashedDataNs(ccName, collConfig.Name))
+				retNamespaces = append(retNamespaces, derivePvtDataNs(ccName, collConfig.Name))
+			}
+		}
+	}
+
+	// iterate each mspID to get implicit collection name
+	// and then add namespaces for each implicit collection per chaincode and new lifecycle namespace)
+	for _, mspID := range mspIDs {
+		implicitCollName := p.GenerateImplicitCollectionForOrg(mspID).Name
+		for _, ccInfo := range chaincodesInfo {
+			retNamespaces = append(retNamespaces, deriveHashedDataNs(ccInfo.Name, implicitCollName))
+			retNamespaces = append(retNamespaces, derivePvtDataNs(ccInfo.Name, implicitCollName))
+		}
+		retNamespaces = append(retNamespaces, deriveHashedDataNs("_lifecycle", implicitCollName))
+		retNamespaces = append(retNamespaces, derivePvtDataNs("_lifecycle", implicitCollName))
+	}
+
+	// add special namespaces: "" and lifecycle management namespaces (new and legacy)
+	retNamespaces = append(retNamespaces, "")
+	retNamespaces = append(retNamespaces, p.Namespaces()...)
+
+	return retNamespaces, nil
+}
+
+// simpleQueryExecutor implements ledger.SimpleQueryExecutor interface
+type simpleQueryExecutor struct {
+	vdb statedb.VersionedDB
```

Comment:
if you embed `statedb.VersionedDB`, you don't need to implement first and third functions below.

Reply:
Changed to embed statedb.VersionedDB. But 1st and 3rd functions are needed because they overwrite the same functions in VersionedDB.